### PR TITLE
Tech/refactor pre/render-service

### PR DESCRIPTION
**Pull request by alschmut**
# Tech/refactor render service

## Description

- Apply height, labels and invert_height separately to settings
- Do not render full map when changing height (scaling) and access existing codemap (slightly better performance for large maps)
- In `PreRenderService`
    - Add `renderData.map`
    - Add `renderData.fileMeta`
    - Remove renderData.renderFile
- Rename `NodeDecorator.decorateFile()` to decorateMap
- Update parameters for `NodeDecorator.decorateFile()` to not use renderFile
- Rename event to `RENDER_MAP_CHANGED_EVENT`

EDIT:
- Extract `renderService.getSortedNodes()`
- Refactor and add `scale()` method in `threeSceneService`
- Refactor `arrowService` to not use `Node.parent` attribute
- Refactor squarify method with more meaningful names
- `TreeMapGenerator` creates unlinked list of nodes
- Remove `Node.children` attribute
- Remove `Node.parent` attribute
- Added tests for refactorService

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [ ] **All** requirements mentioned in the issue are implemented
- [ ] Does match the Code of Conduct and the Contribution file
- [ ] Task has its own **GitHub issue** (something it is solving)
  - [ ] Issue number is **included in the commit messages** for traceability
- [ ] **Update the README.md** with any changes/additions made
- [ ] **Update the CHANGELOG.md** with any changes/additions made
- [ ] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [ ] **All tests pass**
- [ ] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [ ] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [ ] **Manual testing** did not fail

### CODE REVIEW

File: visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts [hunk 1]
```diff
@@ -48,13 +48,7 @@ describe("codeMapRenderService", () => {
...
-	describe("getMapMesh", () => {
-		it("should return _mapMesh via getter", () => {
-			codeMapRenderService["_mapMesh"] = codeMapMesh
-
-			const result = codeMapRenderService.mapMesh
-
-			expect(result).toEqual(codeMapMesh)
-		})
+	describe("something", () => {
+		it("should do something", () => {})
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts]: .

File: visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts [hunk 2]
```diff
@@ -171,11 +198,11 @@ export class CodeMapPreRenderService implements SettingsServiceSubscriber, FileS
 	}
 
 	private notifyFileChanged() {
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: notifyMapChanged

File: visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts [hunk 3]
```diff
@@ -151,6 +169,15 @@ export class CodeMapPreRenderService implements SettingsServiceSubscriber, FileS
 		}
 	}
 
+	private scaleMapIfRenderObjectIsComplete() {
+		if (this.allNecessaryRenderDataAvailable()) {
+			const s: Settings = this.lastRender.settings
+			this.codeMapRenderService.scaleMap(s.appSettings.scaling, s.treeMapSettings.mapSize)
+			this.notifyLoadingMapStatus()
+			this.notifyFileChanged()
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: file didnt change though

File: visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts [hunk 4]
```diff
@@ -58,28 +60,32 @@ export class CodeMapPreRenderService implements SettingsServiceSubscriber, FileS
...
 
 		if (this.lastRender.fileStates && update.fileSettings && (update.fileSettings.blacklist || update.fileSettings.markedPackages)) {
-			this.lastRender.renderFile = this.getSelectedFilesAsUnifiedMap(this.lastRender.fileStates)
-			this.lastRender.renderFile.settings.fileSettings = settings.fileSettings
+			this.updateRenderMapAndFileMeta()
 			this.decorateIfPossible()
 		}
-		this.renderIfRenderObjectIsComplete()
+
+		if (this.settingsOnlyContainNewScaling(update)) {
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: new event onScalingChanged
settingsService aggregates this information to state
consider renaming scale to height in code

File: visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts (hunk 1 above)
Comment (alschmut) [reply visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts]: Tests added

File: visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts (hunk 4 above)
Comment (alschmut) [reply visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: I think creating a separate event should be considered in another issue

**Comment by alschmut** [comment]
I just updated the PR description

File: visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts [hunk 5]
```diff
@@ -58,52 +60,68 @@ export class CodeMapPreRenderService implements SettingsServiceSubscriber, FileS
...
-			this.lastRender.renderFile = this.getSelectedFilesAsUnifiedMap(this.lastRender.fileStates)
-			this.lastRender.renderFile.settings.fileSettings = settings.fileSettings
+			this.updateRenderMapAndFileMeta()
 			this.decorateIfPossible()
 		}
-		this.renderIfRenderObjectIsComplete()
+
+		if (this.allNecessaryRenderDataAvailable()) {
+			if (this.settingsOnlyContainNewScaling(update)) {
+				this.callScaleMap()
```
Comment (NearW) [inline visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: why `CALLScaleMap`? I'd just call it `scaleMap`

Comment (alschmut) [reply visualization/app/codeCharta/ui/codeMap/codeMap.preRender.service.ts]: I renamed it to `scaleMapAndNotify()`

**Comment by NearW** [comment]
Hiding a node and switching modes will result in blacklist being cleared

**Comment by NearW** [comment]
Current master behavior:
- Exclude node in singlemode on MapA
- Exclude node in singlemode on MapB
- Switch to multiple (both nodes are gone now in multiple as well. Blacklist has two entries)
- Switch back to single (MapA will be shown with ALL nodes, blacklist still has two entries)
- Exclude the same node again on MapA as done in step 1
- Blacklist will have 3 entries now, see screenshot

Notice that the path of the 3rd excluded node is not correct at all
Screen Shot 2019-07-12 at 11 02 27

We should merge this and implement a working solution with #476 because the behaviour on master is buggy as well

**Pull request merged**